File: ref/code/L01/div7.cpp
#include <bits/stdc++.h>
using namespace std;

int first[7];
int last[7];
    
int main()
{
	freopen("div7.in", "r", stdin);
	freopen("div7.out", "w", stdout);

	ios::sync_with_stdio(false);
    cin.tie(0), cout.tie(0);

    int n;
    cin>>n;

    memset(first, 0x7f, sizeof(first));
    first[0]=0;
    int prefix = 0;
	for(int i = 1; i <= n; i++) {
		int id;
		cin>>id;
		prefix += id;
		prefix %= 7;
		first[prefix] = min(first[prefix], i);
		last[prefix] = i;
	}
	int ans = 0;
	for(int i = 0; i < 7; i++) {
		ans = max(ans, last[i] - first[i]);
	}
	cout<<ans<<endl;

    return 0;
}

File: ref/code/L01/div7.java
package com.deltacoding.usaco.s2015_2016.jan2016.silver;

import java.io.*;
import java.util.*;

public class div7 {
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new FileReader("div7.in"));
		PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("div7.out")));

		int n = Integer.parseInt(in.readLine());
		int[] first = new int[7];
		int[] last = new int[7];
		Arrays.fill(first, Integer.MAX_VALUE);
		int prefix = 0;
		for(int i = 1; i <= n; i++) {
			prefix += Integer.parseInt(in.readLine());
			prefix %= 7;  // prefix = prefix % 7;
			first[prefix] = Math.min(first[prefix], i);
			last[prefix] = i;
		}
		int ans = 0;
		for(int i = 0; i < 7; i++) {
			ans = Math.max(ans, last[i] - first[i]);
		}
		out.println(ans);
		in.close();
		out.close();
	}
}
